architecture/2025-04-compute-platform-extensibility.md

```diff
@@ -0,0 +1,505 @@
+1. Layered Rendering
+    * Renderers are organized in a layered pattern using composition:
+   KubernetesCompute: kubernetesmetadata → manualscale → daprextension → container
+   ACICompute: aci_manualscale → aci
+    * Each layer handles a specific aspect of rendering, with inner renderers focusing on core functionality and outer renderers adding specialized features
+
+#### Key Type Relationships in the Multiplexer Pattern
+
+1. Core Interface Hierarchy
+    * `renderers.Renderer` is the central interface that defines the rendering contract
+    * All concrete renderers implement this interface, providing a consistent API
+1. Multiplexer Implementation
+    * `mux.Renderer` implements the `renderers.Renderer` interface but delegates to inner renderers
+    * It maintains a map of inner renderers keyed by `EnvironmentComputeKind` (Kubernetes or ACI)
+    * During rendering, it examines the environment compute kind and delegates to the appropriate renderer.
+1. Concrete Renderer Implementations
+    * Platform-specific renderers like `aci.Renderer` and `container.Renderer` (Kubernetes) implement the same interface
+    * Resource-specific renderers like `gateway.Renderer` and `aci_gateway.Renderer` handle specific resource types
+1. Environment Configuration
+    * `EnvironmentCompute` defines the compute environment with its `Kind` property
+    * The multiplexer uses this to determine which inner renderer to invoke
+1. Input and Output Types
+    * `renderers.RenderOptions` provides the context for rendering, including environment configuration
+    * `renderers.RendererOutput` represents the standardized output of the rendering process
+
+This pattern enables the system to:
+
+1. Handle multiple compute environments through a single interface
+1. Dynamically select the appropriate renderer at runtime
+1. Share common interfaces across different platform implementations
+1. Easily extend the system with new compute environments by adding renderers to the map
+
+The design follows the Strategy Pattern, where the rendering strategy is selected based on the environment configuration, while maintaining a consistent interface for the rest of the system.
+
+### Options Considered
+
+#### Proposed Option: Two Milestones
+
+This option has two concurrent milestones. The first is focused on near-term goals. The second is the more traditional design --> build process.
+
+* Milestone 1: Build upon the demo code
+* Milestone 2: Concurrently design an extensibility solution
+
+##### Advantages
+
+* We will have a customer-ready demo faster.
+* Higher-quality design informed by the demo work
+
+##### Disadvantages
+
+* Longer calendar schedule
+
+#### Option 2: Design then build
+
+This option is a traditional development cycle of design --> build.
+
+* Design an extensibility solution.
+* Implement the solution.
+
+### API design (if applicable)
+
+No changes from current design.
```

Comment:
Not sure if it's part of Milestone 1 or Milestone 2, but should there be considerations for how some of the environment and container schemas might change to accommodate compute extensibility? e.g. env compute properties for ACI currently is `kind: 'aci'` with `resourceGroup` required, while for k8s is `kind: 'kubernetes'` with `namespace` required. Would we want to evaluate whether their respective required properties can be collapsed into a since field that encompasses both `namespace` and `resourceGroup` whose value would just be dependent on the target environment?

Reply:
Just a callout, but we also might need to reevaluate how we handle namespace mapping with Kubernetes. We might run into issues with a namespace not being needed with an ACI environment

Reply:
Adding a note here to confirm that working with @willtsai and @sk593 we will be working together on a small design to represent schema changes for milestone 1, which will include namespace mapping.